README.md
# node-red-contrib-moment

**Please note that this node is unlikely to recieve further enhancements now that moment.js is built into JSONata and so is available from change, function and other nodes.**

[Node-RED](http://nodered.org) node ***moment*** produces a nicely formatted Date/Time string using the Moment.JS library. The node is fully time zone/DST/locale aware.

Node ***humanizer*** converts time durations (time spans) into textual descriptions (e.g. 2 minutes).

Both nodes are locale aware regarding the language of the output strings.

![node-appearance](images/node-appearance.png "Node appearance")  
**Fig. 1:** Node appearance


[![Platform](https://img.shields.io/badge/platform-Node--RED-red)](https://nodered.org)
[![NPM Version](https://img.shields.io/npm/v/node-red-contrib-moment.svg?logo=npm)](https://www.npmjs.com/package/node-red-contrib-moment)
[![NPM Total Downloads](https://img.shields.io/npm/dt/node-red-contrib-moment.svg)](https://www.npmjs.com/package/node-red-contrib-moment)
[![NPM Downloads per month](https://img.shields.io/npm/dm/node-red-contrib-moment.svg)](https://www.npmjs.com/package/node-red-contrib-moment)
[![GitHub last commit](https://img.shields.io/github/last-commit/totallyinformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment)
[![GitHub stars](https://img.shields.io/github/stars/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment/watchers)
[![GitHub watchers](https://img.shields.io/github/watchers/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment/stargazers)
[![GitHub license](https://img.shields.io/github/license/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment/blob/master/LICENSE)
[![Min Node Version](https://img.shields.io/node/v/node-red-contrib-moment.svg)](https://www.npmjs.com/package/node-red-contrib-moment)
[![Package Quality](http://npm.packagequality.com/shield/node-red-contrib-moment.png)](http://packagequality.com/#?package=node-red-contrib-moment)
[![Dependencies](https://img.shields.io/david/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment)
[![Open Issues](https://img.shields.io/github/issues-raw/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment/issues)
[![Closed Issues](https://img.shields.io/github/issues-closed-raw/TotallyInformation/node-red-contrib-moment.svg)](https://github.com/TotallyInformation/node-red-contrib-moment/issues?q=is%3Aissue+is%3Aclosed)


Based on thoughts from a [conversation in the Node-RED Google Group](https://groups.google.com/d/msg/node-red/SXEGvfFLfQA/fhJCGBWvYEAJ). Updated with timezone/locale capabilities after Jaques44's initial work. Updated with +/- adjustments after [another conversion in the Google Group](https://groups.google.com/forum/#!topic/node-red/u3qoISFoKus).


<a name="installation"></a>
# Installation

<a name="installation_in_node-red"></a>
## In Node-RED (preferred)

* Via Manage Palette -> Search for "node-red-contrib-moment"

<a name="installation_in_a_shell"></a>
## In a shell
Basic installation:
* go to the Node-RED "userDir" folder, typically `~/.node-red`
* run `npm install node-red-contrib-moment`

To get the latest development version, install with:
* run `npm install TotallyInformation/node-red-contrib-moment`


<a name="moment_usage"></a>
# Usage of node *moment*
The easiest usage of the node is feeding it with an timestamp inject:

![moment-basic-usage](images/moment-basic-usage.png "Node moment usage")  

**Fig. 2:** Basic usage node *moment*


# Updates

Please see the [CHANGELOG](CHANGELOG.md) document.

# Usage

## Moment

The node generally expects an input from the incoming msg. By default, this is msg.payload. If it is a recognisable date/time, it will apply a format and output the resulting string or object accordingly.

Input and output time zones are settable as is the output locale. All of which default to the host systems tz/locale.

This allows the node to be used to translate from one time zone to another. It also will take into account daylight savings time (DST).

You can also apply an adjustment to the date/time by adding or subtracting an amount.

<a name="moment_node_configuration"></a>
## Configuration of node *moment*

![moment-node-settings](images/moment-node-configuration.png "Node moment properties")  

**Fig. 3:** Properties of node *moment*


### *Input from* and *Output to*

These two configuration properties define the `msg` properties in which the input and output data are read from resp. written to. Default is `msg.payload`.

### *Input Timezone* and *Output Timezone*

#### *Input Timezone*

This property defines the timezone of the time fed via the input `msg`. Internally the input time is converted into UTC for further processing.  
The format of *Input Timezone* is in the format *region/location*, e.g. Europe/London. See also timezone lists e.g. built in to [moment-timezone](https://momentjs.com/timezone/) or given in [wikipedia](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones).

**Note**: Spellings are not validated, if it doesn't seem to work, check the validity of *region/location* with these timezone lists.  

The following behaviour is valid:

- If the input data contains a Node-RED timestamp this property is ommitted
  - If the host system has a local timezone set (e.g. `dpkg-reconfigure tzdata` on Linux), the input timestamp is related to this local timezone.
  - If the host system has no local timezone set, the input timestamp is related to UTC.
- If the input data contains an interpretable string, this property is used (to convert internally to UTC).

If left blank in settings, this field may be set from the incoming `msg.inTz` property.

#### *Output Timezone*

This property defines the timezone of the time emitted via the output `msg`.
The format of *Output Timezone* is described above (see *Input Timezone*).

The following behaviour is valid:
- If *Output Format* is left blank, the output format is in 'Zulu' format, independent of the contents of the additional properties *Output Timezone* and *Locale*.

  Zulu format see: https://momentjs.com/docs/#/displaying/as-iso-string/  
  (Example: 2013-02-04T22:44:30.652Z)

If left blank in settings, this field may be set from the incoming `msg.outTz` property.

### *Adjustment*

Using this property, the time can be adjusted by a manually given value. Adjustments can be positive or negative and can be given in milliseconds, seconds, minutes, hours, days, weeks, months, quarters, years.


### *Output Format* and *Locale*

These two properties in combination define the output format emitted in the output `msg`.  

#### *Output Format*

The *Output Format* property defines the format and is described in the [moment.js displaying format section](https://momentjs.com/docs/#/displaying/format/).  

It may be any format string defined by moment.js. The formatting additions from *moment-timezone* are also allowed.  In addition, further (not case sensitive, alternatives in brackets) format strings are also allowed.  

Note that with the exception of ISO8601, other formats are in the specified timezone & DST. If not specified, the output timezone/DST is the same as the input.

Use an output timezone of UTC to force output to that.


##### Format string defined by moment.js

The format string defined by moment.js basically has two options:

- ***Manual given format string:*** This is a string where the time/date parts are represented by characters. Also text parts are allowed. Examples:
  
  - "DD.MM.YYYY HH:mm" gives *20.09.2020 08:30*
  - "dddd, MMMM Do YYYY, h:mm:ss a" gives *Sunday, February 14th 2010, 3:25:50 pm*
  - "[Today is] dddd" gives *Today is Sunday*
  - "[Date: ]YYYY-MM-DD  [Time: ]HH:mm:ss" gives *Date: 2020-09-20  Time: 08:31:45*
  - ["fromNow"](https://momentjs.com/docs/#/displaying/fromnow/) gives *in a month*	
  - ["calendar"](https://momentjs.com/docs/#/displaying/calendar-time/) gives *Last Monday*
  
- ***Predefined localized string:*** This is a string which defines a localized format. Examples:
  
  - "LLL" gives *October 20, 2020 8:33 AM*
  - "LTS" gives *8:30:25 PM*
  - "llll" gives *Thu, Sep 4, 1986 8:30 PM*

For more options see https://momentjs.com/docs/#/displaying/format/.

##### Format string is left blank resp. is "ISO8601" or "ISO"  

In this case the output is in ISO 8602 format, e.g. "2015-01-28T16:24:48.123Z".

Note that ISO8601 formatted output is ALWAYS in UTC ('Z', Zulu time) not local, no matter what output timezone you may specify.

See also [`moment().toISOString()`](https://momentjs.com/docs/#/displaying/as-iso-string/).

##### Format string is "date" resp. "jsDate"

This is a *Javascript Date object* in the form `{years:nnnn, months:n, date:n, hours:n, minutes:n, seconds:n, milliseconds:n}`. 

It may be used for manual (fixed) data/time values.

WARNING: moment.js has a bizarre object format where the month is zero-based (0-11) instead of 1-based (1-12) like all the other elements are. I don't currently know why, I've raised an upstream issue but this appears to be a deliberate decision for some strange reason. 

See also [`moment().toObject()`](https://momentjs.com/docs/#/displaying/as-object/).

##### Format string is "fromNow" resp. "timeAgo"

This is a human readable output, e.g. *30 minutes ago* or *in a month* (only rough time spans are given in this output format type, see also the *humanizer* example below). The time span is derived from the actual time and the time fed into the node.

See also [`moment().fromNow()`](https://momentjs.com/docs/#/displaying/fromnow/).

##### Format string is "calendar" resp. "aroundNow"

This is a human readable alternative, e.g. *Last Monday* or *Tomorrow 2:30pm*.
Note that dates beyond a week from now are output as yyyy-mm-dd.

See also [`moment().calendar()`](https://momentjs.com/docs/#/displaying/calendar-time/).

##### Format string is "date" resp. "jsDate"

This output format type is actually not working (see [issue #37](https://github.com/TotallyInformation/node-red-contrib-moment/issues/37)).


#### *Locale*

In case of a textual output string contents the *Locale* property defines the language of the textual parts (e.g. "October" vs. "Oktober" vs. "ottobre" vs. "lokakuuta").

If the output is shown in the wrong format, such as dates in US mm/dd/yy format, change the output locale. For example, using en_gb will force short dates to output in dd/mm/yy format. The default is en which moment assumes means the USA :-(

See also [Locale Helper](https://lh.2xlibre.net/locales/) (Note: Not every locale given there is supported).


### *Topic* (additional topic)

Using this property you can add an additional topic to the output `msg.topic`.
A resulting `msg` may be (value "myTopicString"):

```json
{"_msgid":"b16b00b5.bada55","payload":"2020-09-20T12:47:55.143Z","topic":"myTopicString"}
```


## Input of node *moment*

Input values in the object **Input from** can be of the following types:

- ***timestamp:*** The current date/time is used as input.
- ***msg***, ***global*** or ***flow*** and the given property is empty or does not exist: The current date/time is used as input.
- ***JSON date time object:*** This data time object may contain the following elements: *years*, *months*, *days*, *hours*, *minutes*, *seconds*, *milliseconds*.
  
  Example: `{"years":2020,"months":1,"date":11,"hours":5,"minutes":6}`.

  If elements are not given (e.g. *years* and *months* are missing in the object) the actual time values are used instead.

- ***a property containing a string that is a recognizable date/time:*** The value will be interpreted and processed.
  
  Example: `2020-02-11T05:06`

- ***a property containing a numeric value:*** The value will be assumed to be a [UNIX time value](https://momentjs.com/docs/#/displaying/unix-timestamp-milliseconds/) (ms since 1970-01-01). Remark: This is the format which the node *Inject* emits at option **timestamp**.
  
- ***a property containing a string that is not a recognisable date/time (including `null`):*** Then no conversion takes place, the output will be an empty string plus a debug warning.


Note that parsing date/time strings is a hard problem. moment.parseFormat helps but it isn't magic. We assume that ambiguous input dates such as 3/5/05 are in EU/UK format dd/mm/yy unless either the input timezone is in America or the locale is set to en_US.


## Outputs of node *moment*

If the **output** property is not `msg.payload` the input `msg.payload` is retained in the output.

The date/time output is a formatted string if the configuration property ***Output Format*** is anything other than *date* resp. *jsDate* or *object* in which case the output is a Javascript date object or an object as described below respectively.

Output string formatting is controlled by the ***Locale*** and the ***Output Format*** setting. Note that the output Timezone is ignored for ISO8601 output (the default), such output is always in UTC. For other formats, the output will be in the specified timezone which defaults to your host timezone.

Specifying different input and output timezones allows you to translated between them.

The output `msg` will pass through the input `msg.topic` unless it is overridden by the ***Topic*** configuration property. If the ***Output to*** field is changed from the default `msg.payload`, the input `msg.payload` will also be passed through.



# Usage of node *humanizer*

This node converts an input time span to a humanized text string to the output `msg.payload.humanized`. The language of the output string is derived from the locale of the system, i.e. it is not changeable (like the *Locale* property of the *moment* node).  

See also [moment.duration().humanize()](https://momentjs.com/docs/#/durations/humanize/).

(Contributed by [Laro88](https://github.com/Laro88))


## Configuration of node *humanizer*

![humanizer-node-settings](images/humanizer-node-configuration.png "Node humanizer properties")  
**Fig. 4:** Properties of node *humanizer*

### 'Input variable'

This property defines the input `msg.payload` property which shall be used for the conversion. If left blank, `msg.payload` is used.

## Input of node *humanizer*

The input is a number which defines a time span in seconds.

## Outputs of node *humanizer*

The output is a string object in `msg.payload.humanized`.  
The time spans are evaluated in intervals, see *humanizer* example for details.


# Examples

***

CHANGELOG.md
Please note that this node is unlikely to recieve further enhancements now that moment.js is built into JSONata and so is available from change, function and other nodes.

- 5.0.0 - Update dependencies, replace `os-locale` with `os-locale-s`.

  os-locale library is now ESM only which won't work with Node-RED. It is also bloated. Replaced with the lighter-weight os-locale-s library

  **WARNING** Neither os-locale nor os-locale-s seems to work correctly (at least on Windows 11) - it is returning `en_US` rather than `en_GB` on my UK only W11 system.

  **BREAKING** Minimum node.js version is now 14.14 and the minimum Node-RED version is v3.

  **FIXED** - output format `x` and `X` now accepted since the moment.js library is now updated.

- 4.0.0 - Dependency update, output settings, bug fix
   - **BREAKING CHANGE** Updated dependencies mean that minimum Node.js version is now v10 (previously v8). Minimum Node-RED version is now v1.0 (prevously v0.13)
   - Update dependencies
   - Move change log to separate file from readme.
   - Contributed changes, see [Issue #34](https://github.com/TotallyInformation/node-red-contrib-moment/issues/34):
     - Explanation extended, examples added
     - Testing flows moved from subdir examples\ to subdir testing\
     - .html help extended  
   - Prep for Node-RED v1.
   - Workaround for upstream parseFormat bug, see [Issue #24](https://github.com/TotallyInformation/node-red-contrib-moment/issues/24). If input is an ISO date string with seconds >3dp, the input is trimmed to 3dp seconds.
   - Added `msg.settings` to output msg. Shows the settings/parameters used. Allows for further processing or validation if needed.
   - Added ability to override input and output Timezone settings from an incoming msg object. The appropriate setting must be empty and the properties to use are `msg.inTz` and `msg.outTz`. Resolves [Issue #31](https://github.com/TotallyInformation/node-red-contrib-moment/issues/31).
   - Bug fix, locale now correctly defaults to host locale if not provided.
  
- 3.0.3 - Just a version number bump to fix npm.
- 3.0.2 - Update dependencies to latest versions.
- 3.0.1 - **Bug Fix** Remove errant log statement.
- 3.0.0 - No feature changes, just upgrades of the dependent libraries. Note breaking change, minimum Node.JS version is now v6 or above.
    - Documentation updated to clarify processing of different inputs.
    - Fix: If input property contains `null`, output is now a warning + empty string in line with other invalid inputs. It was an incorrect timestamp.
    - Examples added to Node-RED library to make testing easier.
- 2.0.7 - Fallback to new Date(inp) when the date string is non-standard, like this output from the file stat() function: `2017-12-12 16:03:51.832427000 -0400` as well as the Twitter API default time string: `Mon Jan 08 21:24:37 +0000 2018` (contributed by [Steve Rickus](https://github.com/shrickus)). Update dependencies.
- 2.0.6 - Upstream change in MomentJS introduced bug when feeding with a timestamp (number), fixed
- 2.0.5 - Autocorrect common tz errors (e.g. UTC+4 should be ETC/GMT+4) & autofill default
- 2.0.4 - Show warning for invalid timezone specifications
- 2.0.3 - Humanize helper node added, exposes moments humanize of timespan functionality (contributed by [Laro88](https://github.com/Laro88)),
  Documentation updated to highlight issue with moment.js object handing (month is 0-11 not 1-12)
- 2.0.2 - Minor fixes - change version to get latest moment.js. 2017-03-19
- 2.0.1 - Fix get/set of msg properties. 2016-07-08
- 2.0.0 - Significant rewrite, updated moment.js, got rid of all eval's, added adjustment calcs, added time zone and locale awareness. 2016-06-26
- 1.0.9 - Merged in some fixes on Jacques44's contributions & acknowledged him in the package. Also fixed the npm readme. 2016-06-12
- 1.0.5 - Merged a pull request containing a Locale option for localisation. 2016-03-30
- 1.0.3 - First stable release. 2015-01-31


moment/nrmoment.js
/**
 * Copyright (c) 2020 Julian Knight (Totally Information)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

// Node for Node-Red that outputs a nicely formatted string from a date/time
// object or string using the moment.js library.
'use strict'

// require moment.js (must be installed from package.js as a dependency)
const moment = require('moment-timezone')
const parseFormat = require('moment-parseformat')
// const osLocale = require('os-locale')
const { osLocale } = require('os-locale-s')
const hostTz = moment.tz.guess()
const hostLocale = osLocale.sync()


// Module name must match this nodes html file
const moduleName = 'moment'

/** Catch & correct input strings in ISO format with seconds >3dp
 * since parseFormat has an unfixed bug (Issue #24, https://github.com/gr2m/moment-parseformat/issues/96)
 * @param {string} inp Input date/time string
 */
function catchDp(inp) {
    // If ISO string w/ seconds >3dp, then process else exit
    if ( inp.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{4,}/) ) {
        inp = inp.replace(/^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3})\d*(.*)/, '$1$2')
    }

    return inp
}

/** Export the function that defines the node */
module.exports = function (RED) {
    'use strict'

    // The main node definition - most things happen in here
    function nodeDefinition(config) {
        // Create a RED node
        RED.nodes.createNode(this, config)

        // Store local copies of the node configuration (as defined in the .html)
        this.topic = config.topic
        this.input = config.input || 'payload' // where to take the input from
        this.inputType = config.inputType || 'msg' // msg, flow, global, timestamp or string
        this.fakeUTC = config.fakeUTC || false // is the input UTC rather than local date/time?
        this.adjAmount = config.adjAmount || 0 // number
        this.adjType = config.adjType || 'days' // days, hours, etc.
        this.adjDir = config.adjDir || 'add' // add or subtract
        this.format = config.format || '' // valid moment.js format string
        this.locale = config.locale || hostLocale // valid moment.js locale string - default to host locale (v4)
        this.output = config.output || 'payload' // where to put the output
        this.outputType = config.outputType || 'msg' // msg, flow or global
        this.inTz = config.inTz || false // timezone, '' or zone name, e.g. Europe/London
        this.outTz = config.outTz || this.inTz // timezone, '' or zone name, e.g. Europe/London

        /** copy "this" object in case we need it in context of callbacks of other functions. */
        var node = this

        // respond to inputs....
        /** @since v3.1.0, amended ready for Node-RED v1 */
        node.on('input', function (msg, send, done) {
            'use strict' // We will be using eval() so lets get a bit of safety using strict

            // If this is pre-1.0, 'send' will be undefined, so fallback to node.send
            send = send || node.send
            // If this is pre-1.0, 'done' will be undefined, so fallback to dummy function
            done = done || function(){}

            
            // If the node's topic is set, copy to output msg
            if (node.topic !== '') {
                msg.topic = node.topic
            } // If nodes topic is blank, the input msg.topic is already there

            // make sure output property is set, if not, assume msg.payload
            if (node.output === '') {
                node.output = 'payload'
                //node.warn('Output field is REQUIRED, currently blank, set to payload');
            }
            if (node.outputType === '') {
                node.outputType = 'msg'
                node.warn('Output Type field is REQUIRED, currently blank, set to msg')
            }

            /** If the input property is blank, assume NOW as the required timestamp
             * or make sure that the node's input property actually exists on the input msg
             * @type {(string|Date|null)}
             **/
            var inp = ''
            // If input is a blank string, use a Date object with Now DT
            if (node.input === '') {
                inp = new Date()
            } else {
                // Otherwise, check which input type & get the input
                try {
                    switch (node.inputType) {
                        case 'msg':
                            inp = RED.util.getMessageProperty(msg, node.input)
                            break
                        case 'flow':
                            inp = node.context().flow.get(node.input)
                            break
                        case 'global':
                            inp = node.context().global.get(node.input)
                            break
                        case 'date':
                            inp = new Date()
                            break
                        case 'str':
                            inp = node.input.trim()
                            break
                        default:
                            inp = new Date()
                            node.warn('Unrecognised Input Type, ' + node.inputType + '. Output has been set to NOW.')
                    }
                } catch (err) {
                    inp = new Date()
                    node.warn('Input property, ' + node.inputType + '.' + node.input + ', does NOT exist. Output has been set to NOW.')
                }
            }
            // We are going to overwrite the output property without warning or permission!

            // Final check for input being a string (which moment doesn't really want to handle)
            // NB: moment.js v3 will stop accepting strings. v2.7+ throws a warning.
            /** @type {(string|Object)} */
            var dtHack = ''
            /** @type {(string|Date)} */
            var inpFmt = ''
            // @from v3 2018-09-23: If input is `null`, change to empty string
            if (inp === null) inp = ''
            if ((typeof inp) === 'string') {
                inp = inp.trim()
                // Some string input hacks
                switch (inp.toLowerCase()) {
                    case 'today':
                        inp = new Date()
                        break
                    case 'yesterday':
                        inp = new Date()
                        dtHack = { days: -1 }
                        break
                    case 'tomorrow':
                        inp = new Date()
                        dtHack = { days: +1 }
                        break
                    default:
                        // Catch string ISO inputs with seconds >3dp which parseFormat treats incorrectly (Issue #24) - trim to 3dp only
                        inp = catchDp(inp)
                        // Try to guess locale so as to guess whether DMY or MDY
                        var prefOrder = { preferredOrder: { '/': 'DMY', '.': 'DMY', '-': 'YMD' } }
                        if ((node.locale.toLowerCase().replace('-', '_') === 'en_US') || (node.inTz.split('/')[0] === 'America')) {
                            prefOrder = { preferredOrder: { '/': 'MDY', '.': 'DMY', '-': 'YMD' } }
                        }
                        inpFmt = parseFormat(inp, prefOrder)
                }
            } else if ((typeof inp) === 'number') {
                // @from 2017-06-15 IF inp is a number at this point, it needs turning into a date object
                inp = new Date(inp)
            }

            // At this point, `inp` SHOULD be a Date object and safe to pass to moment.js

            // Check if input TZ is provided in the msg?
            if (msg.inTz) {
                node.inTz = msg.inTz
            } else {
                // If the input TZ is not set ...
                if (!node.inTz) {
                    // take it from the host TZ
                    node.inTz = hostTz 
                }
            }
            // Check if output TZ is provided in the msg?
            if (msg.outTz) {
                node.outTz = msg.outTz
            } else {
                // If the output TZ is not set ...
                if (!node.outTz) { 
                    // make it the same as the input TZ
                    node.outTz = node.inTz 
                }
            }

            // Validate input and output timezones - @since 2.0.4
            // TODO: Needs more validation
            if (moment.tz.zone(node.inTz) === null) {
                // tz invalid, warn and set to UTC
                node.warn('Moment: Input Timezone Invalid, reset to UTC - see http://momentjs.com/timezone/docs/#/data-loading/')
                node.inTz = 'UTC'
            }
            if (moment.tz.zone(node.outTz) === null) {
                // tz invalid, warn and set to UTC
                node.warn('Moment: Output Timezone Invalid, reset to UTC - see http://momentjs.com/timezone/docs/#/data-loading/')
                node.outTz = 'UTC'
            }

            // Get a Moment.JS date/time - NB: the result might not be
            //  valid since the input might not parse as a date/time
            var mDT
            if (inpFmt !== '') {
                mDT = moment.tz(inp, inpFmt, true, node.inTz)
            } else {
                // @from 2017-06-15 change to momentjs meant having to add null parameter
                mDT = moment.tz(inp, null, true, node.inTz)
            }

            // Fallback to JS built-in Date parsing, if not recognized by moment
            if (!mDT.isValid()) {
                var dtm = new Date(inp)
                if (dtm === 'Invalid Date') {
                    node.warn('Unrecognized date string format => ' + inp)
                }
                else {
                    mDT = moment(dtm)
                }
            }

            // Adjust the date for input hacks if needed (e.g. input was "yesterday" or "tommorow")
            if (dtHack !== '') { mDT.add(dtHack) }

            // JK: Added OS locale lookup
            // JK: Add a trap to Jaques44's locale code in case the output locale string is invalid
            try {
                // Jacques44 - set locale for localised output formats
                mDT.locale(node.locale)
            } catch (err) {
                node.warn('Locale string invalid - check moment.js for valid strings')
            }

            // Adjust the input date if required
            if (node.adjAmount !== 0) {
                // check if measure is valid
                if (isMeasureValid(node.adjType)) {
                    // NB: moments are mutable so don't need to reassign
                    if (node.adjDir === 'subtract') {
                        mDT.subtract(node.adjAmount, node.adjType)
                    } else {
                        mDT.add(node.adjAmount, node.adjType)
                    }
                } else {
                    // it isn't valid so warn and don't adjust
                    node.warn('Adjustment measure type not valid, no adjustment made - check moment.js docs for valid measures (days, hours, etc)')
                }
            }

            // ==== NO MORE DATE/TIME CALCULATIONS AFTER HERE ==== //

            // If required, change to the output Timezone
            if (node.outTz !== '') mDT.tz(node.outTz)

            // Check if the input is a date?
            if (!mDT.isValid()) {
                // THIS SHOULD NEVER BE CALLED - it is left to catch the occasional error
                node.warn('The input property was NOT a recognisable date. Output will be a blank string')
                setOutput(msg, node.outputType, node.output, '')
            } else {
                // Handle different format strings. We allow any fmt str that
                // Moment.JS supports but also some special formats

                // If format not set, assume ISO8601 string if input is a Date otherwise assume Date
                switch (node.format.toLowerCase()) {
                    case '':
                    case 'iso8601':
                    case 'iso':
                        // Default to ISO8601 string
                        setOutput(msg, node.outputType, node.output, mDT.toISOString())
                        break
                    case 'fromnow':
                    case 'timeago':
                        // We are also going to handle time-from-now (AKA time ago) format
                        setOutput(msg, node.outputType, node.output, mDT.fromNow())
                        break
                    case 'calendar':
                    case 'aroundnow':
                        // We are also going to handle calendar format (AKA around now)
                        // Force dates >1 week from now to be in ISO instead of US format
                        setOutput(msg, node.outputType, node.output, mDT.calendar(null, { sameElse: 'YYYY-MM-DD' }))
                        break
                    case 'date':
                    case 'jsdate':
                        // we also allow output as a Javascript Date object
                        setOutput(msg, node.outputType, node.output, mDT.toDate())
                        break
                    case 'object':
                        // we also allow output as a Javascript Date object
                        setOutput(msg, node.outputType, node.output, mDT.toObject())
                        break
                    default:
                        // or we assume it is a valid format definition ...
                        setOutput(msg, node.outputType, node.output, mDT.format(node.format))
                }
            }

            // Include settings in output
            msg.settings = {
                'input': inp,
                'input_format': inpFmt,
                'input_tz': node.inTz,
                'output_format': node.format,
                'output_locale': node.locale,
                'output_tz': node.outTz,
            }

            // Send the output message
            send(msg)
            // Finished processing input msg (NR 1+)
            done()
        })

        // Tidy up if we need to
        //node.on('close', function() {
        // Called when the node is shutdown - eg on redeploy.
        // Allows ports to be closed, connections dropped etc.
        // eg: node.client.disconnect();
        //});

        // Set the appropriate output variable
        function setOutput(msg, outputType, output, value) {
            try {
                switch (outputType) {
                    case 'msg':
                        RED.util.setMessageProperty(msg, output, value)
                        break
                    case 'flow':
                        node.context().flow.set(output, value)
                        break
                    case 'global':
                        node.context().global.set(output, value)
                        break
                    default:
                        node.warn('Unrecognised Output Type, ' + outputType + '. No output.')
                }
            } catch (err) {
                node.warn('Output property, ' + outputType + '.' + output + ', cannot be set. No output.', err)
            }
        } // --- end of setOutput function --- //

        // Is the date/time adjustment type (measure) valid? See moment.js docs
        function isMeasureValid(adjType) {
            var validTypes = ['years', 'y', 'quarters', 'Q', 'months', 'M', 'weeks', 'w', 'days', 'd', 'hours', 'h', 'minutes', 'm', 'seconds', 's', 'milliseconds', 'ms']
            //return validTypes.includes(adjType);
            return validTypes.indexOf(adjType) > -1
        } // --- end of isMeasureValid function --- //

    } // ---- end of nodeGo function ---- //

    /** Register the node by name. This must be called before overriding any of the Node functions. */
    RED.nodes.registerType(moduleName, nodeDefinition)

    // Create API listener: sends the host locale & timezone to the admin ui
    // NB: uses Express middleware on the admin server
    RED.httpAdmin.get('/contribapi/moment', RED.auth.needsPermission('moment.read'), function (req, res) {
        res.json({
            'tz': hostTz,
            'locale': hostLocale
        })
    })

    /** Lookup API for country and zone names
     * TODO: Not yet completed and no front-end settings yet
     */
    RED.httpAdmin.get('/contribapi/momentzones', RED.auth.needsPermission('moment.read'), function (req, res) {
        if ( req.query.country ) {
            console.log('[nrmoment:get:zones] Country URL param provided', req.query.country)
        }
        res.json(moment.tz.names())
    })
}
